[2022-04-01-preview] Add New Api-version for Microsoft.ApiManagement

solankisamir opened the pull request:
## ARM API Information (Control Plane)

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow.
</i>
Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

solankisamir committed:
Adds base for updating Microsoft.ApiManagement from version preview/2021-12-01-preview to version 2022-04-01-preview

solankisamir committed:
Updates readme

solankisamir committed:
Updates API version in new specs and examples

Hardell committed:
APIM Auth Servers (#19234)

* APIM Auth Servers

* adding x-ms-identifiers

* removing some weird, invisible special char

* formatting

* oAuth2AuthenticationSettings moved to AuthSettings

* Formatting

Co-authored-by: Milan Zolota <mizolota@microsoft.com>

seandkim committed:
API Management Authorization Endpoints (#19615)

* Add blockchain to latest profile

* Add additional types

* add authorizations definitions

* authorizations operations

* add examples

* update readme

* fix examples

* fix linter delete errors

* address CI validation errors

* prettier fix

* update to 2022-04

* fix readme

* Update specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimauthorizationproviders.json

Co-authored-by: Sean Kim <seaki@microsoft.com>

* update versions

* Apply suggestions from code review

Co-authored-by: Mark Cowlishaw <markcowl@microsoft.com>
Co-authored-by: Annaji Sharma Ganti <anganti@microsoft.com>
Co-authored-by: Annaji Sharma Ganti <42851022+annaji-msft@users.noreply.github.com>

solankisamir committed:
Move Long running Create Operation from Location based to Azure-AsyncOperation Header (#19733)

* azure-asyncOperation

* prettier

tomkerkhove committed:
fix(apim): Add missing 'metrics' property to diagnostics contract in 2022-04-01-preview (#20317)

DeagleGross committed:
apim /PUT apis import add translateRequiredQueryParameters (#20333)

rkmanda commented:
>         "count": {

this is not the typical implementation of a collection response in ARM. Please remove

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:1133 in 5256a61. [](commit_id = 5256a61c505929673ce5df347205f1d84cef9be3, deletion_comment = False)

rkmanda commented:
>           "$ref": "./../../../../../common-types/resource-management/v2/types.json#/definitions/Resource"

use either trackedresource or proxyresource instead

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:1268 in 5256a61. [](commit_id = 5256a61c505929673ce5df347205f1d84cef9be3, deletion_comment = False)

rkmanda commented:
>         "count": {

remove

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:1313 in 5256a61. [](commit_id = 5256a61c505929673ce5df347205f1d84cef9be3, deletion_comment = False)

solankisamir commented:
> > ```
> >       "$ref": "./../../../../../common-types/resource-management/v2/types.json#/definitions/Resource"
> > ```
> 
> use either trackedresource or proxyresource instead
> 
> Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:1268 in [5256a61](https://github.com/Azure/azure-rest-api-specs/commit/5256a61c505929673ce5df347205f1d84cef9be3). [](commit_id = [5256a61](https://github.com/Azure/azure-rest-api-specs/commit/5256a61c505929673ce5df347205f1d84cef9be3), deletion_comment = False)

@rkmanda I have opened a new PR request 
https://github.com/Azure/azure-rest-api-specs/pull/20461

honghr commented:
/azp run unifiedPipeline

raych1 commented:
@solankisamir , you said this version is only deployed to canary region, will this version roll out to other regions soon?

solankisamir committed:
[2022-04-01-preview] Replace resource with proxyresource and TrackedResource (#20461)

* replace resource with proxyresource

* revert to proxyresource

solankisamir commented:
> @solankisamir , you said this version is only deployed to canary region, will this version roll out to other regions soon?

Yes, we will start rolling out this version to all regions, once we merge the PR to the main branch.

LFZ96 committed:
Add type object to authorization definitions (#20631)

Authorization definitions were missing "type": "object", and this change adds that key/value pair

Mielek committed:
Add type object to policy fragment definition (#20585)

malincrist committed:
APIM Open ID Connect providers (#20622)

* APIM Open ID Connect providers

* added new proeprties for update

* prettier

solankisamir commented:
/azp run

solankisamir commented:
@ArcturusZhang how should I resolve this merge conflict? I tried to create this PR, but it is huge

https://github.com/Azure/azure-rest-api-specs/pull/20860

The Change came because of your PR https://github.com/Azure/azure-rest-api-specs/pull/20756 and since `apim-2022-01-01-preview` is common branch, I can't directly push to it.

mentat9 commented:
@solankisamir - Does this comment (https://github.com/Azure/azure-rest-api-specs/pull/20399#issuecomment-1231032119) mean ARM review is no longer required for this PR? If so, please either Close this PR or remove the WaitForARMReview label.

GuanchenIntern committed:
[APIM] Add Nat Gateway (#19990)

* Update apimdeployment.json

* Create ApiManagementCreateServiceWithNatGatewayEnabled.json

* fix typo in file

* Change Nat Gateway property to enum

* modify type of natgateway state

* update property name

* add example reference

* small fix in example

* rename to  outboundPublicIPAddresses

Co-authored-by: Samir Solanki <samirsolanki@outlook.com>

solankisamir commented:
/azp run

solankisamir committed:
Merge branch 'main' into apim-2022-04-01-preview

VatsaPatel committed:
[2022-04-01-preview] MIgrate2Stv2 API (#20504)

* migrate2stv2

* updated to post

* 202 and location

* add body to 202

* remove body from 202

Co-authored-by: Vatsa Patel <vatsapatel@microsoft.com>
Co-authored-by: Samir Solanki <samirsolanki@outlook.com>
Co-authored-by: vatsapatel@microsoft.colm <vatsapatel@microsoft.colm>

solankisamir commented:
@msyyc can we merge this PR if things look good?

LFZ96 committed:
Address Authorizations MissingTypeObject errors (#20919)

LFZ96 committed:
Add forgotten If-Match header (#20920)

* Add forgotten If-Match header

`If-Match` header for the `DeleteAuthorizationAccessPolicy.json file` was forgotten. This change adds the wildcard character for the `If-Match` header for that file.

* Update ApiManagementDeleteAuthorization.json

msyyc commented:
Hi @solankisamir Why did you remove these labels:
![image](https://user-images.githubusercontent.com/70930885/192974052-2c650c4d-70b3-4e4b-b565-ba1a5702a796.png)

msyyc commented on specification/apimanagement/resource-manager/readme.md:
```diff
@@ -31,12 +31,64 @@ openapi-type: arm
 tag: package-2021-08
```
@solankisamir  Please change the default tag to your new tag otherwise SDK pipeline can't check the tag you added.

solankisamir replied:
@msyyc We don't release SDK for preview API-versions. We will be creating a stable api-version after this and then we will update this tag.

msyyc replied:
> @msyyc We don't release SDK for preview API-versions. We will be creating a stable api-version after this and then we will update this tag.

@solankisamir Now that the new preview tag will be released for SDK, please review the breaking. @qiaozha @ArcturusZhang

solankisamir commented:
@msyyc a gentle reminder of this!

mentat9 commented:
> > Hi @solankisamir Why did you remove these labels: ![image](https://user-images.githubusercontent.com/70930885/192974052-2c650c4d-70b3-4e4b-b565-ba1a5702a796.png)
> 
> Each commit to this api-version is already ARM Signed Off and I also saw this comment
> 
> [#20399 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/20399#issuecomment-1258756438)

@solankisamir - I think you must have misunderstood my comment. I wasn't telling you ARM signoff isn't required. I was responding to your comment that you had opened another PR. I expected you would close this PR to replace it with the other PR, but that's not the case: you are moving forward with this one. This PR is adding a new API version with new APIs, so ARM signoff is required. I've added back the appropriate labels. Current oncall will help you complete ARM review.

TimLovellSmith commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimanagement.json:
```diff
@@ -0,0 +1,824 @@
+      },
+      "description": "Error Field contract."
+    },
+    "ErrorResponse": {
```
>ErrorResponse [](http://example.com/codeflow?start=5&length=13)

I would not define it here, and instead directly reference

 "$ref": "../../../../../common-types/resource-management/v3/types.json#/definitions/ErrorResponse"

everywhere needed

TimLovellSmith commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimanagement.json:
```diff
@@ -0,0 +1,857 @@
+        },
+        "count": {
+          "type": "integer",
+          "format": "int64",
```
>format [](http://example.com/codeflow?start=11&length=6)

consider marking as readonly true

TimLovellSmith replied:
Or even better, just remove it! Its not normal to include this in ARM collection apis.

TimLovellSmith commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimanagement.json:
```diff
@@ -0,0 +1,857 @@
+      "required": true,
+      "type": "string",
+      "description": "Version of the API to be used with the client request."
+    },
```
I would not define it here, and instead directly reference

 "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/ApiVersionParameter"

everywhere needed

TimLovellSmith commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimanagement.json:
```diff
@@ -0,0 +1,857 @@
+      "maxLength": 256,
+      "pattern": "^[^*#&+:<>?]+$",
+      "x-ms-parameter-location": "method"
+    },
```
What are the format or nature of these id parameters - GUID? Resource ID? Something else?

TimLovellSmith commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimanagement.json:
```diff
@@ -0,0 +1,857 @@
+      "pattern": "^[^*#&+:<>?]+$",
+      "x-ms-parameter-location": "method"
+    },
+    "ResourceGroupNameParameter": {
```
>ResourceGroupNameParameter [](http://example.com/codeflow?start=5&length=26)

Referencing common-types will simplify reviewing

TimLovellSmith commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimanagement.json:
```diff
@@ -0,0 +1,857 @@
+      "pattern": "^[^*#&+:<>?]+$",
+      "x-ms-parameter-location": "method"
+    },
+    "SubscriptionIdParameter": {
```
>SubscriptionIdParameter [](http://example.com/codeflow?start=5&length=23)

Referencing common-types will simplify reviewing

TimLovellSmith commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimauthorizationproviders.json:
```diff
@@ -0,0 +1,805 @@
+        "x-ms-odata": "./definitions.json#/definitions/AuthorizationProviderContract"
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/authorizationProviders/{authorizationProviderId}": {
```
>authorizationProviderId [](http://example.com/codeflow?start=152&length=23)

Just wondering, is it really an ID or a name?

solankisamir replied:
@LFZ96 can you comment?

LFZ96 replied:
It's a name

TimLovellSmith replied:
Want to update the parameter naming to match that statement then?

LFZ96 replied:
@TimLovellSmith Yes, I will check with my team. I'm new to the project and the person that has worked on this document has left the team. I remember having the same question and him giving an explanation for it. I will report back.

LFZ96 replied:
@TimLovellSmith My understanding is that it's the identifier of the authorization provider that coincides with the name that the user gives it. For example, `google-provider-123`. It's unique and is used to identify the auth provider rather than something like a GUID. These entities are also close 1-to-1s of Connector team connectors/connections where the names of those are similarly `ConnectorId` and `ConnectionId`.

TimLovellSmith commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimauthorizationproviders.json:
```diff
@@ -0,0 +1,805 @@
+          }
+        ],
+        "responses": {
+          "200": {
```
>" [](http://example.com/codeflow?start=10&length=1)

Do you only support Update? What about Create?? Create of a tracked resource should be able to return 201 Created in addition to 200 OK.

solankisamir replied:
@LFZ96 can you take a look?

LFZ96 replied:
@TimLovellSmith We support Create AuthorizationProvider
Should I change this to a 201 response status instead of the 200?

TimLovellSmith replied:
I'm thinking you probably want both 200 and 201 responses for all your PUT operations.

solankisamir replied:
@LFZ96 this is guideline @TimLovellSmith  is referring to

[https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc015-put-apis-that-only-return-200](https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc015-put-apis-that-only-return-200)

TimLovellSmith commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimdeployment.json:
```diff
@@ -0,0 +1,1955 @@
+          "ApiManagementService"
+        ],
+        "operationId": "ApiManagementService_MigrateToStv2",
+        "description": "Upgrades an API Management service to the Stv2 platform. This change is not reversible. This is long running operation and could take several minutes to complete.",
```
>the Stv2 platform. [](http://example.com/codeflow?start=62&length=18)

Will this description be meaningful to people reading the doc, if they don't have context?

solankisamir replied:
added more details

TimLovellSmith replied:
Great link!

TimLovellSmith commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimdeployment.json:
```diff
@@ -0,0 +1,1955 @@
+          },
+          "readOnly": true,
+          "description": "Outbound public IP address prefixes associated with NAT Gateway deployed service. Available only for Premium SKU."
+        },
```
Want to provide more info? Will this support IPv6?

TimLovellSmith commented:
>         "count": {

consider adding reaodnly

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:1279 in 5b71742. [](commit_id = 5b71742e4b41c01a194c4014e56dab3459d4d8e2, deletion_comment = False)

TimLovellSmith commented:
>       },

Are these Azure AD? It seems implied. But you could be explicit.

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:1318 in 5b71742. [](commit_id = 5b71742e4b41c01a194c4014e56dab3459d4d8e2, deletion_comment = False)

TimLovellSmith commented:
>           "description": "The post redirect url after login is completed."

Is this 'POST' as in the HTTP verb, or 'post-redirect' as in 'after redirect'?

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:1326 in 5b71742. [](commit_id = 5b71742e4b41c01a194c4014e56dab3459d4d8e2, deletion_comment = False)

TimLovellSmith commented:
>           "description": "The post redirect url after login is completed."

URL

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:1326 in 5b71742. [](commit_id = 5b71742e4b41c01a194c4014e56dab3459d4d8e2, deletion_comment = False)

TimLovellSmith commented:
>         "count": {

Or even better, just remove it! Its not normal to include this in ARM collection apis.

---
In reply to: [1268715259](https://github.com/Azure/azure-rest-api-specs/pull/20399#issuecomment-1268715259) [](http://example.com/codeflow?ancestors=1268715259)

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:1279 in f60e9cf. [](commit_id = f60e9cfc0d9fec05f768bb01d90b3a8848e2e06c, deletion_comment = False)

TimLovellSmith commented:
>           "description": "Total record count number across all pages."

Are you even going to know the total count across all pages? In general, ARM paged APIs might distribute reqeusts among multiple endpoints, and then filter the results based on RBAC permissions - so the count in the returned response might not match the collections actually seen by the user.

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:3714 in 5b71742. [](commit_id = 5b71742e4b41c01a194c4014e56dab3459d4d8e2, deletion_comment = False)

TimLovellSmith commented:
>           "description": "If true, the Open ID Connect provider may be used in the developer portal test console. True by default if no value is provided."

Any security impact of this setting? If so, document it clearly! Is this the secure by default option?

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:3814 in 5b71742. [](commit_id = 5b71742e4b41c01a194c4014e56dab3459d4d8e2, deletion_comment = False)

TimLovellSmith reviewed (changes requested):
:clock1:

solankisamir commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimauthorizationproviders.json:
```diff
@@ -0,0 +1,805 @@
+          }
+        ],
+        "responses": {
+          "200": {
```
@LFZ96 can you check this comment

LFZ96 replied:
@TimLovellSmith Similar to above, we do support Creation. Should I change the status to be 201 instead of 200?

TimLovellSmith replied:
Create operation is generally really 'create or update' and can either/both 200 and 201.

JeffreyRichter replied:
A side note: PUT is generally create or replace and PATCH is create or update. In HTTP, a PUT over an existing resource is the logical equivalent of a DELETE/PUT. We're trying to make this (PUT=create or replace) be the culture in Azure (since it is the culture for HTTP). It matters a lot because of the way Azure does its versioning.

solankisamir commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimdeployment.json:
```diff
@@ -0,0 +1,1955 @@
+        "x-ms-long-running-operation": true,
+        "x-ms-long-running-operation-options": {
+          "final-state-via": "location"
+        }
```
the CI system raises alerts if we dont provide it

solankisamir commented:
>         "count": {

This is some requirement from UI team. We are not able to remove it and have got exception from ARM Review board. This is only used for Proxy Resource, which are bound to a single region.

Updated readonly property.

---
In reply to: [1268721756](https://github.com/Azure/azure-rest-api-specs/pull/20399#issuecomment-1268721756) [](http://example.com/codeflow?ancestors=1268721756,1268715259)

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:1279 in 5b71742. [](commit_id = 5b71742e4b41c01a194c4014e56dab3459d4d8e2, deletion_comment = False)

solankisamir commented:
>           "description": "The post redirect url after login is completed."

Thanks updated everywhere

---
In reply to: [1268717124](https://github.com/Azure/azure-rest-api-specs/pull/20399#issuecomment-1268717124) [](http://example.com/codeflow?ancestors=1268717124,1268717124)

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:1326 in 5b71742. [](commit_id = 5b71742e4b41c01a194c4014e56dab3459d4d8e2, deletion_comment = False)

solankisamir commented:
>           "description": "Total record count number across all pages."

This is only for Proxy Resource which are bound to a single APIM service in a single region. 

---
In reply to: [1268722888](https://github.com/Azure/azure-rest-api-specs/pull/20399#issuecomment-1268722888) [](http://example.com/codeflow?ancestors=1268722888,1268722888)

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:3714 in f60e9cf. [](commit_id = f60e9cfc0d9fec05f768bb01d90b3a8848e2e06c, deletion_comment = False)

TimLovellSmith commented:
>                 "description": "The policy document is not Xml encoded and is hosted on a http endpoint accessible from the API Management service."

XML

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:4262 in 5b71742. [](commit_id = 5b71742e4b41c01a194c4014e56dab3459d4d8e2, deletion_comment = False)

TimLovellSmith commented:
>           "description": "Binary OR value of the Snippet scope.",

Does it need capitalizing in the middle of a sentence? I'm thinking probably not...

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:4299 in 5b71742. [](commit_id = 5b71742e4b41c01a194c4014e56dab3459d4d8e2, deletion_comment = False)

TimLovellSmith commented on custom-words.txt:
```diff
@@ -16,6 +16,7 @@ ACCELCONTAINER
 acceptors
 accesskey
 accesspoint
+accesspolicies
```
You shouldn't need to add this. You should instead be using PascalCase, CamelCase versions of it, or just using plain english documentation "access polices"

TimLovellSmith replied:
Still true, I think?

TimLovellSmith replied:
(and nonbreaking if so)

TimLovellSmith commented:
>           "description": "Descriptions of APIM policies."

Inconsideration of corporate documentation style...

Despite this being in the context of API Management service, APIM may not be clear to all readers, and style guide is to avoid excessive acronyms. [Consider that your provider name is ApiManagement, and not APIM, and your general documentation doesn't use the acronym: [https://learn.microsoft.com/en-us/azure/api-management/](https://learn.microsoft.com/en-us/azure/api-management/)]

Please align this with your public facing documentation and replace APIM acronym with "API management" or "API Management" (your standard capitalization for your documentation) in descriptions everywhere.

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:4312 in 5b71742. [](commit_id = 5b71742e4b41c01a194c4014e56dab3459d4d8e2, deletion_comment = False)

solankisamir commented:
>     "ArmIdWrapper": {

We got this object from WebApp Private Endpoint Spec. We can consider this as feedback for next stable api-version change.

---
In reply to: [1268729460](https://github.com/Azure/azure-rest-api-specs/pull/20399#issuecomment-1268729460) [](http://example.com/codeflow?ancestors=1268729460,1268729460)

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:7434 in 5b71742. [](commit_id = 5b71742e4b41c01a194c4014e56dab3459d4d8e2, deletion_comment = False)

solankisamir commented:
>           "description": "If true, the Open ID Connect provider may be used in the developer portal test console. True by default if no value is provided."

`"description": "If true, the Open ID Connect provider may be used in the developer portal test console. True by default if no value is provided."`

No security impact. Customer configures the Open ID Connect Provider and we provide them a TEST Console, where can they test out stuff. This has been the By Default feature.

With property customer can choose for this to not be used in Test Console.

---
In reply to: [1268724748](https://github.com/Azure/azure-rest-api-specs/pull/20399#issuecomment-1268724748) [](http://example.com/codeflow?ancestors=1268724748,1268724748)

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:3814 in f60e9cf. [](commit_id = f60e9cfc0d9fec05f768bb01d90b3a8848e2e06c, deletion_comment = False)

solankisamir committed:
Use common types for specs and count as readonly (#21023)

* common types

* count readonly

solankisamir commented:
>           "description": "Binary OR value of the Snippet scope.",

It is literally the `OR` operator

---
In reply to: [1270361942](https://github.com/Azure/azure-rest-api-specs/pull/20399#issuecomment-1270361942) [](http://example.com/codeflow?ancestors=1270361942,1270361942)

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:4299 in 5b71742. [](commit_id = 5b71742e4b41c01a194c4014e56dab3459d4d8e2, deletion_comment = False)

solankisamir committed:
Sasolank/more review comments (#21025)

* XML

* proxy to gateway

LFZ96 commented:
> > ```
> >   },
> > ```
> 
> Are these Azure AD? It seems implied. But you could be explicit.
> 
> Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:1318 in [5b71742](https://github.com/Azure/azure-rest-api-specs/commit/5b71742e4b41c01a194c4014e56dab3459d4d8e2). [](commit_id = [5b71742](https://github.com/Azure/azure-rest-api-specs/commit/5b71742e4b41c01a194c4014e56dab3459d4d8e2), deletion_comment = False)

@TimLovellSmith I think they are Azure AD. Can you give me example of how to be more explicit?

TimLovellSmith commented:
> > > ```
> > >   },
> > > ```
> > 
> > 
> > Are these Azure AD? It seems implied. But you could be explicit.
> > Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:1318 in [5b71742](https://github.com/Azure/azure-rest-api-specs/commit/5b71742e4b41c01a194c4014e56dab3459d4d8e2). [](commit_id = [5b71742](https://github.com/Azure/azure-rest-api-specs/commit/5b71742e4b41c01a194c4014e56dab3459d4d8e2), deletion_comment = False)
> 
> @TimLovellSmith I think they are Azure AD. Can you give me example of how to be more explicit?

I think it was just extra confusing for me in the context of 'application programming interface management', and I wanted to be sure you don't mean some other kind of application id.

LFZ96 commented:
> > ```
> >       "description": "The post redirect url after login is completed."
> > ```
> 
> Is this 'POST' as in the HTTP verb, or 'post-redirect' as in 'after redirect'? Or just 'post-login' as in after login?
> 
> Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:1326 in [f60e9cf](https://github.com/Azure/azure-rest-api-specs/commit/f60e9cfc0d9fec05f768bb01d90b3a8848e2e06c). [](commit_id = [f60e9cf](https://github.com/Azure/azure-rest-api-specs/commit/f60e9cfc0d9fec05f768bb01d90b3a8848e2e06c), deletion_comment = False)

@TimLovellSmith  After talking with my team, it's supposed to be the redirect URL _after_ the login has been successfully completed. I will update the description.

msyyc commented:
Hi @JeffreyRichter Could you help review the swagger breaking

JeffreyRichter commented:
Why are these breaks being introduced? Was the swagger just always wrong and this PR fixes it?

LFZ96 committed:
Update Authorizations Spec (#21027)

* Update definitions.json

Update wording for PostGetLoginLink endpoint description

* Update apimauthorizationproviders.json

Add 201 response to all Authorization PUT requests

* Updated examples and fixed formatting

There was a formatting issue within apimauthorizationproviders.json, and the Authorization examples needed to be updated with the new 201 responses for creating/updating Authorization entities.

* Add long-running-operation key/value

Added x-ms-long-running-operation: true to Authorization PUT requests

* Remove long-running-operations

solankisamir committed:
readonly revert (#21050)

agapovm committed:
Set  SchemaContract.Document as required. (#20110)

* Updated documentation of the SchemaContract. Server use to return code 500 in case SchemaContract.Document is null. That issue was fixed in the APIM and server will return proper response code.

* Fix AzureApiValidation

* update field with properties

solankisamir commented:
> Why are these breaks being introduced? Was the swagger just always wrong and this PR fixes it?

@JeffreyRichter @mikekistler we have reverted the `readonly` property for all resource collection response.

TimLovellSmith commented:
>         "id": "/subscriptions/subid/resourceGroups/rg1/providers/Microsoft.ApiManagement/service/apimService1/authorizationProviders/aadwithauthcode/authorizations/authz1/accesspolicies",

-> camelCase "accessPolicies"

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/examples/ApiManagementCreateAuthorizationAccessPolicy.json:20 in a317f91. [](commit_id = a317f91afaf965c4c6820c35f3301469fa591546, deletion_comment = False)

TimLovellSmith commented:
>         "id": "/subscriptions/subid/resourceGroups/rg1/providers/Microsoft.ApiManagement/service/apimService1/authorizationProviders/aadwithauthcode/authorizations/authz1/accesspolicies",

-> camelCase "accessPolicies"

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/examples/ApiManagementCreateAuthorizationAccessPolicy.json:31 in a317f91. [](commit_id = a317f91afaf965c4c6820c35f3301469fa591546, deletion_comment = False)

solankisamir committed:
revert remaining readonly on collection (#21051)

solankisamir commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimauthorizationproviders.json:
```diff
@@ -0,0 +1,841 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/authorizationProviders/{authorizationProviderId}/authorizations/{authorizationId}/accesspolicies": {
```
@LFZ96 can you please address this?

solankisamir replied:
@TimLovellSmith this is corrected. Please review again

solankisamir commented:
@mikekistler here are details of the Breaking change in this PR.

<html>
<body>
<!--StartFragment-->

❌ 1007 - RemovedClientParameter | The new version is missing a client parameter that was found in the old version. Was 'ApiVersionParameter' removed or renamed?New: Microsoft.ApiManagement/preview/2022-04-01-preview/apimanagement.json#L132:3Old: Microsoft.ApiManagement/stable/2021-08-01/apimanagement.json#L132:3

>> We moved to common types and this is now showing as Breaking Change. We did it on recommendation of @TimLovellSmith 

❌ 1007 - RemovedClientParameter | The new version is missing a client parameter that was found in the old version. Was 'ResourceGroupNameParameter' removed or renamed?New: Microsoft.ApiManagement/preview/2022-04-01-preview/apimanagement.json#L132:3Old: Microsoft.ApiManagement/stable/2021-08-01/apimanagement.json#L132:3

>> We moved to common types and this is now showing as Breaking Change. We did it on recommendation of @TimLovellSmith 

❌ 1007 - RemovedClientParameter | The new version is missing a client parameter that was found in the old version. Was 'SubscriptionIdParameter' removed or renamed?New: Microsoft.ApiManagement/preview/2022-04-01-preview/apimanagement.json#L132:3Old: Microsoft.ApiManagement/stable/2021-08-01/apimanagement.json#L132:3

>> We moved to common types and this is now showing as Breaking Change. We did it on recommendation of @TimLovellSmith 

❌ 1034 - AddedRequiredProperty | The new version has new required property 'document' that was not found in the old version.New: Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json#L5724:9Old: Microsoft.ApiManagement/stable/2021-08-01/definitions.json#L5059:9

>> This is approved breaking change here https://github.com/Azure/azure-rest-api-specs/pull/20110

❌ 1034 - AddedRequiredProperty | The new version has new required property 'document' that was not found in the old version.New: Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json#L5737:5Old: Microsoft.ApiManagement/stable/2021-08-01/definitions.json#L5072:5

>> This is approved breaking change here https://github.com/Azure/azure-rest-api-specs/pull/20110

❌ 1034 - AddedRequiredProperty | The new version has new required property 'document' that was not found in the old version.New: Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json#L5724:9Old: Microsoft.ApiManagement/stable/2021-08-01/definitions.json#L5059:9

>> This is approved breaking change here https://github.com/Azure/azure-rest-api-specs/pull/20110

❌ 1034 - AddedRequiredProperty | The new version has new required property 'document' that was not found in the old version.New: Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json#L5737:5Old: Microsoft.ApiManagement/stable/2021-08-01/definitions.json#L5072:5

>> This is approved breaking change here https://github.com/Azure/azure-rest-api-specs/pull/20110

<!--EndFragment-->

mikekistler commented:
What about

<img width="836" alt="image" src="https://user-images.githubusercontent.com/85643503/195164543-5a5a0f39-28f2-4197-9af2-4926fd1f7f4f.png">

This appears to change "ApiContract" from a "Resource" to a "ProxyResource". Is this intentional? Shouldn't this be considered a breaking change?

solankisamir commented:
> What about
> 
> <img alt="image" width="836" src="https://user-images.githubusercontent.com/85643503/195164543-5a5a0f39-28f2-4197-9af2-4926fd1f7f4f.png">
> 
> This appears to change "ApiContract" from a "Resource" to a "ProxyResource". Is this intentional? Shouldn't this be considered a breaking change?

@mikekistler 
The `Resource` construct was incorrect. These are all `ProxyResources` in ARM, so this is correction of that definition.

LFZ96 committed:
Change to camel casing for "accesspolicies" (#21070)

* Change to camel casing for "accesspolicies"

* More camel casing updates for access policies

solankisamir committed:
list example fixed (#21089)

solankisamir commented:
/azp run

raosuhas commented:
>             "$ref": "../../../../../common-types/resource-management/v2/types.json#/parameters/ApiVersionParameter"

v3 is the preferred version here. Any reason you decided to update to an older version ?

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimapis.json:87 in 8b7c808. [](commit_id = 8b7c808b4f0af427e64ad789603de67f2fd3f33c, deletion_comment = False)

raosuhas commented:
>             "$ref": "../../../../../common-types/resource-management/v2/types.json#/parameters/ApiVersionParameter"

Same comment across this PR

---
In reply to: [1276850247](https://github.com/Azure/azure-rest-api-specs/pull/20399#issuecomment-1276850247) [](http://example.com/codeflow?ancestors=1276850247)

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimapis.json:87 in 7169f4f. [](commit_id = 7169f4fb5cd1185eef0c977b5d151f0c36c55d9d, deletion_comment = False)

raosuhas commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimauthorizationproviders.json:
```diff
@@ -0,0 +1,841 @@
+            "description": "Authorization was successfully deleted."
+          },
+          "204": {
+            "description": "The authorization server settings were successfully deleted."
```
>successfully [](http://example.com/codeflow?start=67&length=12)

nit : no content. the resource was not found

raosuhas commented:
> > ```
> >       "$ref": "./../../../../../common-types/resource-management/v2/types.json#/definitions/Resource"
> > ```
> 
> use either trackedresource or proxyresource instead
> 
> In reply to: [1229125657](https://github.com/Azure/azure-rest-api-specs/pull/20399#issuecomment-1229125657)
> 
> In reply to: [1229125657](https://github.com/Azure/azure-rest-api-specs/pull/20399#issuecomment-1229125657)
> 
> Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:1268 in [5256a61](https://github.com/Azure/azure-rest-api-specs/commit/5256a61c505929673ce5df347205f1d84cef9be3). [](commit_id = [5256a61](https://github.com/Azure/azure-rest-api-specs/commit/5256a61c505929673ce5df347205f1d84cef9be3), deletion_comment = False)

Please fix this : This is explained in the guidelines here : 

https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#collections

<a name="collections-avoid-count-property> ⚠️ YOU SHOULD NOT return a count of all objects in the collection as this may be expensive to compute.

raosuhas commented:
@solankisamir Looks mostly good. 

Please consider fixing the comment on the count in your collection responses since it is not recommended to return it and also updating to v3 if possible.

solankisamir committed:
fix definition (#21110)

solankisamir committed:
upgrade to v3 for common types (#21109)

* upgrade to v3

* Space

* revert to v2 proxyResource

solankisamir commented:
> > ```
> >         "$ref": "../../../../../common-types/resource-management/v2/types.json#/parameters/ApiVersionParameter"
> > ```
> 
> v3 is the preferred version here. Any reason you decided to update to an older version ?
> 
> Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/apimapis.json:87 in [8b7c808](https://github.com/Azure/azure-rest-api-specs/commit/8b7c808b4f0af427e64ad789603de67f2fd3f33c). [](commit_id = [8b7c808](https://github.com/Azure/azure-rest-api-specs/commit/8b7c808b4f0af427e64ad789603de67f2fd3f33c), deletion_comment = False)

@raosuhas we have moved to v3 wherever possible. Thanks for the pointer

solankisamir commented:
> @solankisamir Looks mostly good.
> 
> Please consider fixing the comment on the count in your collection responses since it is not recommended to return it and also updating to v3 if possible.

@raosuhas I have opened issue https://github.com/Azure/azure-rest-api-specs/issues/21111 which gives details in how we are using the `count` property. Previously we have got a signoff from Vlad, as these are all `proxy` resources in ARM.


If things look good, can you take a second look.

mikekistler commented:
Breaking changes all appear to be false negatives. Marking as approved.

TimLovellSmith commented:
>           "type": "string",

Consider adding format arm-id?

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:7440 in 46f9565. [](commit_id = 46f9565353ee46e3c7372be0b7d938debd67fcc9, deletion_comment = False)

TimLovellSmith commented:
>                 "description": "The policy document is not XML encoded and is hosted on a http endpoint accessible from the API Management service."

HTTP

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2022-04-01-preview/definitions.json:4262 in 46f9565. [](commit_id = 46f9565353ee46e3c7372be0b7d938debd67fcc9, deletion_comment = False)

solankisamir commented:
Thanks @TimLovellSmith let me work on these changes on separate PR. I don't want to reset Breaking Change Approval status.
Will create the PR, once this is merged. Hopefully none of this is blocking.

msyyc commented:
@solankisamir please resolve the `change requested` otherwise the merge is blocked 
![image](https://user-images.githubusercontent.com/70930885/195741900-e452f48f-3d3d-4493-aec6-d3479d7b5770.png)
